### DISCUSSION TITLE
4D/5D Functionality in BlenderBIM

### DISCUSSION THREAD

**Discussion by vulevukusej** [IfcOpenShell/IfcOpenShell#1779] [Q&A] [answered]
Hey all,

Is there somewhere I can see a development concept/vision for these modules inside of BlenderBIM?  I´ve been testing the BB plugin a lot over the past weeks but I am struggling a little to envisage what this might look like in the future.  When I think of other software that is particularly strong in these aspects (DesiteMD, BexelManager, iTWO), I wonder whether such complexity and functionality is perhaps not a bit too much for the UI in Blender.  Perhaps this is just down to my own experience, as I´ve found trying to create a nice UI for a plugin in Blender quite cumbersome.. 

In any case, I am a heavy user of the abovementioned software's and am happy to share my experience.

Cheers,

Vukas

**Comment by Moult** [comment, marked as answer]
Hey @vulevukusej ! Sorry for the late reply. My focus for the next release is under the hood stability and engine upgrades (new testable architecture, new v0.7.0 branch), so unfortunately, the development of 4D/5D has stalled somewhat.

The vision is indeed similar to what is shown in tools like Bexel Manager et friends - parametric, generated, filtered based application of cost rates, productivity rates, and so on.

A fun fact is that the BlenderBIM Add-on is designed to be quite decoupled from Blender as a whole. The system is designed to support multiple applications and multiple UIs. Ever since we started the 4D/5D stuff, we knew that one day, we'd need to probably build a more purpose-built UI outside Blender. This extends to other modules too, like drawings and sheet management.

The core IfcOpenShell API is completely agnostic of both Blender and the BlenderBIM Add-on. All calls are passed through an API run function. The idea is to create a daemon which speaks to and from the IFC server (which may run within the BlenderBIM Add-on for now), and therefore a browser based UI (or something else, but I reckon a web-based one is a good idea) will be used for more bespoke UIs where Blender's highly prescriptive UI toolkit falls short. This includes gantt charts, hover, drag and drop, and spreadsheet-like interfaces. Changes in one UI will be reflected in the other. This is funky because in the future, this allows users to also integrate into their own UIs, or run headlessly.

Hope it makes sense. We've planned for it and the code architecture allows for it, it's just mostly a question of developer time. Will you be the guy who makes it happen sooner rather than later? :)

**Reply by vulevukusej** [reply]
Appreciate the reply mate!  

What you describe makes absolute sense.  When I first wrote the question I had only just started looking into the BlenderBIM and IfcOpenShell code in detail so concepts such as decoupling and agnostic software made absolutely no sense to me.  Now I´m simply in awe of the work you and the other developers have done so I look forward to making it happen sooner rather than later ;).  

I need to learn how to code competently first (in progress!) but in the meantime I´ll keep promoting this and other OS Tools in the german-speaking world.
